### ISSUE TITLE
progress of a song is remembered even when I skip to another song

### ISSUE THREAD

**Issue by lucamp** [vanilla-music/vanilla#665]
- Vanilla Music Version: 1.0.54
- Android Version: kitkat 4.4.2
- Phone vendor / model: lg d620
- Application was downloaded from: F-Droid

## Description
progress of a song is remembered even when I skip to another song

## Steps to reproduce the issue
Pause a song, wait few minutes and skip to a previous song. When the playlist arrives at the song that was paused, it starts from the position it was left.

This issue started for me after an old vanilla update long time ago but unfortunately it does not happen every time so I am not able to reproduce it.

Vanilla should remember the actual song position just when the song is not changed, So if I skip to another with backward or forward button, or by choosing another in the queue, the position should no longer remembered

**Comment by adrian-bl** [comment]
Hmm, that is interesting, haven't seen this yet.

There are not many places in the code where we do a seek: The only non-user initiated seek is done when we are loading the state in PlaybackService.java:1367

Looking at the code, it *might* be possible that we do an invalid seek if:

* We get paused at song A
* Get put into background, we save the state
* Get started again and read the old state
* Then fail to seek to the old song (or seek before it was initialized)

However: I don't know how likely this is to happen.
When you say 'wait few minutes and skip to a previous song': Do you exit Vanilla Music in this time frame? And how do you skip to the previous song? Using a Widget?

Apart from that: there is also a slim chance that it might be related to the gapless code where we recycle old mediaplayer instances.

**Comment by lucamp** [comment]
I tried for at least one hour but I am still not able to reproduce it with consistency.

Nevertheless I almost use only the widget and when It happens the main screen is not open.

I mean few minutes just because after I pause the music it happens only after my phone has gone to lockscreen automatically. 
-- 
Sent from my Android phone with GMX Mail. Please excuse my brevity.On 06/09/2017, 19:29 Adrian Ulrich <notifications@github.com> wrote:
Hmm, that is interesting, haven't seen this yet.
There are not many places in the code where we do a seek: The only non-user initiated seek is done when we are loading the state in PlaybackService.java:1367
Looking at the code, it might be possible that we do an invalid seek if:

We get paused at song A
Get put into background, we save the state
Get started again and read the old state
Then fail to seek to the old song (or seek before it was initialized)

However: I don't know how likely this is to happen.
When you say 'wait few minutes and skip to a previous song': Do you exit Vanilla Music in this time frame? And how do you skip to the previous song? Using a Widget?
Apart from that: there is also a slim chance that it might be related to the gapless code where we recycle old mediaplayer instances.

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/vanilla-music/vanilla","title":"vanilla-music/vanilla","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/vanilla-music/vanilla"}},"updates":{"snippets":[{"icon":"PERSON","message":"@adrian-bl in #665: Hmm, that is interesting, haven't seen this yet.\r\n\r\nThere are not many places in the code where we do a seek: The only non-user initiated seek is done when we are loading the state in PlaybackService.java:1367\r\n\r\nLooking at the code, it *might* be possible that we do an invalid seek if:\r\n\r\n* We get paused at song A\r\n* Get put into background, we save the state\r\n* Get started again and read the old state\r\n* Then fail to seek to the old song (or seek before it was initialized)\r\n\r\nHowever: I don't know how likely this is to happen.\r\nWhen you say 'wait few minutes and skip to a previous song': Do you exit Vanilla Music in this time frame? And how do you skip to the previous song? Using a Widget?\r\n\r\nApart from that: there is also a slim chance that it might be related to the gapless code where we recycle old mediaplayer instances.\r\n\r\n"}],"action":{"name":"View Issue","url":"https://github.com/vanilla-music/vanilla/issues/665#issuecomment-327556235"}}}

**Comment by adrian-bl** [comment]
Thanks for the info: I actually managed to reproduce this.

There are quite some factors required to hit this issue:

* Play a song, hit pause
* Wait > 60s to have the player enter sleep state (+ save the seek position)
* Let Android kill the Application
* Restart Vanilla again by hitting |< or >|

This will cause us to load the old seek state but play a different song - when we reach the song for which we loaded the state, we will seek to the old position, which is stupid.

Commit https://github.com/vanilla-music/vanilla/commit/65fc2929b67b77338475ed7b90dc9df3cc62ef1b fixes this.
